### PR TITLE
BLS signature integration

### PR DESCRIPTION
**Pull request by david-stolp**

### CODE REVIEW

File: node/client/src/lib.rs [hunk 1]
```diff
@@ -144,13 +145,19 @@ impl Client {
...
+        for (i, authority) in authorities.iter().enumerate() {
+            if authority.account_id == self.signer.account_id {
+                let shard_block_signature = shard_block.sign(&self.signer);
+                shard_block.add_signature(shard_block_signature, i);
+                let block_signature = block.sign(&self.signer);
+                block.add_signature(block_signature, i);
+                break;
+            }
+        }
+        // TODO: Remove.  This is a hack to make tests pass.
```
Comment (ilblackdragon) [inline node/client/src/lib.rs]: also an issue for this. I also not clear what is this hack for.

File: core/primitives/src/types.rs [hunk 2]
```diff
@@ -43,6 +42,45 @@ impl<'a> From<&'a ReadablePublicKey> for PublicKey {
...
+    #[serde(with = "bs58_serializer")]
+    pub signature: BlsSignature,
+    pub authority_mask: AuthorityMask,
+}
+
+impl GroupSignature {
+    // TODO (optimization): It's better to keep the signature in projective coordinates while
+    // building it, then switch to affine coordinates at the end.  For the time being we just keep
+    // it in affine coordinates always.
+    pub fn add_signature(&mut self, signature: &PartialSignature, authority_id: usize) {
```
Comment (ilblackdragon) [inline core/primitives/src/types.rs]: This usize catches an eye here. I see that you use it from .enumerate below, but may be just use u32 here and `as u32` there?

File: core/primitives/src/aggregate_signature.rs [hunk 3]
```diff
@@ -90,17 +110,150 @@ impl<E: Engine> PublicKey<E> {
...
+    }
+}
+
+impl<E: Engine> From<&PublicKey<E>> for Vec<u8> {
+    fn from(key: &PublicKey<E>) -> Self {
+        Self::from(key.compress().as_ref())
+    }
+}
+
+// TODO: TryFrom instead of From since conversion can fail
```
[the remark this answers is not in the document]
Comment (david-stolp) [reply core/primitives/src/aggregate_signature.rs]: #502

File: node/client/src/lib.rs (hunk 1 above)
Comment (david-stolp) [reply node/client/src/lib.rs]: Removed.

**Pull request merged**